### ISSUE TITLE
Cache results from executeOnText() if filename is provided

### ISSUE THREAD

**Issue by IanVS** [eslint/eslint#4607]
Currently, the results of `executeOnText()` are not cached, which makes it impossible for build tools like [gulp-eslint](https://github.com/adametry/gulp-eslint) use eslint caching.  It seems from a cursory look that it might be possible to cache the results as long as a valid (resolvable) filename is provided as the second argument of `executeOnText()`.  

@royriojas, @nzakas, is this something you think might be possible and desirable to add?

**Comment by nzakas** [comment]
Should work in theory, so long as we create and store the hash just like we do with files.

I wonder, though, why gulp-eslint is doing that instead of executeOnFiles?

**Comment by royriojas** [comment]
TL;DR: It is possible but we will have to expose the method to persist the cache.

It is feasible, but I guess we will have to expose a method to persist the cache, something like `CLIEngine.persistCache()`. That's because by default all operations in the cache are done in memory and just before the results are printed the cache is saved to disk. 

So for example if they have an stream of 10 files, they will have to call `CLIEngine.persistCache()` at the end of processing the stream to save the cache.

**Comment by royriojas** [comment]
@nzakas I guess because as part of the gulp philosophy, they load all files and convert them to a stream of files. Those streams are passed to the tasks, so they don't do I/O operations. Each task return the stream with the applied transform.

**Comment by ilyavolodin** [comment]
Not specifically related to the issue at hand, but `persistCache` would be reasonable method to add. In general, I think, the whole caching structure has to be extracted out of CLIEngine and re-written. Right now, the way it's done, is causing a number of issues with refactoring and adding new features (parallel linting and glob configs, for example).

**Comment by nzakas** [comment]
@royriojas ah, of course, thanks for explaining.

@ilyavolodin is parallel linting still coming? :)

**Comment by royriojas** [comment]
@IanVS can you elaborate about the issues, parallel linting should not be an issue, Even keeping the sync calls to determine if the file has changed. Those can be changed as well if needed.

But what is glob configs?

**Comment by sindresorhus** [comment]
:+1: I could use this for [gulp-xo](https://github.com/sindresorhus/gulp-xo) too, which works similar to `gulp-eslint`.

**Comment by ilyavolodin** [comment]
@nzakas Eventually - yes:-) But not for v2:-(

@royriojas Caching isn't blocking parallel linting, it just makes it extremely hard to refactor existing code. Glob configs is an ability to specify different configs based on globs (like different rules for jsx vs js files, for example). Since we do caching per directory, and glob config breaks that assumption, it's blocking glob config feature to a point.

**Comment by royriojas** [comment]
Is parallel linting expecting launch different processes to divide the file list? is the division of the file list predictable?

the cache expects to be in memory until is time to save the results, so running eslint from multiple processes might be an issue. Unless there is a way to pass the cache object to each process and only call reconcile from the top process.

I would like to help to make the cache code not a blocker for the refactorings you're planning so anything that I can help to improve on that side please let me know

**Comment by royriojas** [comment]
by the way if this is accepted I can work on the changes to use the cache from `executeOnText`.

**Comment by IanVS** [comment]
I was hoping you'd say that, @royriojas.  Thanks.  :smile:

**Comment by ilyavolodin** [comment]
@royriojas Yes, parallel linting would launch different processes for each file (or rather create a pool of N processes and reuse them). Caching will be handled in the main thread, as in, if the file is already cached, it will not even going to be passed to the pool. Caching is not really blocking parallel processing, it just makes it very hard to refactor CLIEngine to allow for both async and sync execution. For more information take a look at this issue: #3565

**Comment by royriojas** [comment]
@ilyavolodin will it help if the cache is always handled from the main process? I mean once a linter process finish the linting then it sends a message to the parent process with the files that passed and the ones that don't. we cache the one that passed and remove from the cache (or simply don't store) the ones that failed. Will take a look to #3565 now

**Comment by ilyavolodin** [comment]
@royriojas Yes, that's exactly the way I implemented it. This isn't a problem, problem is that unless I refector cache out of CLIEngine, I need to duplicate a lot of cache logic in the code. Take a look at #3565 I explained there what is causing issues with parallel linting.

**Comment by nzakas** [comment]
Given that parallel linting is still a ways off, let's not make that a blocking consideration for this. We can always say caching won't work with parallel linting to start.

**Comment by IanVS** [comment]
@nzakas does that mean this issue can get the "approved" tag added?

**Comment by nzakas** [comment]
I'm having a philosophically hard time with the idea. `executeOnText` is meant to work on text only, and I'm having a hard time rationalizing that it should be augmented with caching, which is really for working with files.

Right now, we don't do caching for text that is piped in...I need more convincing that this is a good idea.

**Comment by platinumazure** [comment]
For what it's worth, I'm with @nzakas here. If a consuming tool chooses to send in text rather than files, they should do their own caching.

**Comment by IanVS** [comment]
If `executeOnText` is only meant to work on text, why do we allow a second argument to serve as a filename?  It has proven to be very powerful for tools like editor linters and gulp plugins, because it allows `.eslintignore` to be respected, while making it possible to lint streams and "dirty" (unsaved) files.  The `--cache` option is given to the CLIEngine constructor, not directly to the `executeOnText` function, so arguably it's more confusing that caching works in some cases and not in others.  It seems like this is a fairly straightforward feature that @royriojas has agreed to work on, and it will add value to users.

Asking other tools to all implement their own (likely inconsistent) caching behaviors would be taking a step backwards, IMHO.  It's possible to add caching to gulp with [gulp-cache](https://github.com/jgable/gulp-cache), but it is very basic and does not consider whether files pass linting, only whether they are modified (so simply running it twice makes it look like all errors are fixed).

@adametry do you have any thoughts on this issue?  How difficult would it be for you to add your own caching layer around eslint to re-implement its caching functionality?

**Comment by royriojas** [comment]
@IanVS the cache feature in eslint is done by [file-entry-cache](https://www.npmjs.com/package/file-entry-cache). That's separated from eslint, technically `gulp-eslint` can use it directly if this issue is not finally accepted. 

the method [hasFileChanged](https://github.com/royriojas/file-entry-cache/blob/master/cache.js#L29)

is the one that will work better in this scenario. If it is the first time the file is seen it will just return true, otherwise it will return false.

Also if for some reason (like the file failed) you want to not store the file in the cache it can be removed with [removeEntry](https://github.com/royriojas/file-entry-cache/blob/master/cache.js#L142)

At the end of the stream you just need to make sure the `cache.reconcile()` [method](https://github.com/royriojas/file-entry-cache/blob/master/cache.js#L167) is called and you will have it working.

**Comment by nzakas** [comment]
Some history: The optional filename was added so you could get some configuration information applied to the text and was requested for text editor implementations (who didn't want to write text to the disk in order to lint continuously). It doesn't necessarily follow that all features available for file-based linting are also available for text-based linting.

My main concern is that this is actually more hairy than it seems. For instance, outside of a Gulp context, what if there's a file on disk with the same name as a the file passed in for text? The cache is built based on the text, not what's on disk, yet it's persisted to disk in the cache. How can we reconcile that if they then run eslint in a different way?

We've gone down this path before, where it seemed like adding something was simple and ended up a big mess (glob patterns on the command line), and I just want to make sure we're not heading in that direction again.

**Comment by royriojas** [comment]
I have to agree with @nzakas, but on the bright side, as I mentioned, it is fairly easy to integrate file-entry-cache in other projects as well. I can even help `gulp-eslint` with this if required, @IanVS.

**Comment by IanVS** [comment]
OK, that makes sense.  Thanks @royriojas for explaining how it could be done within a consuming tool.  Maybe that is the best approach after all.

**Comment by ilyavolodin** [comment]
Can we just expose caching functionality through cli-engine and let people use it if the want? Might eliminate a problem and take a maintenance burden off our hands?

**Comment by adametry** [comment]
@IanVS, I played around with this idea while developing `gulp-eslint@1.1.0`, prompted by a [feature request](https://github.com/adametry/gulp-eslint/issues/99). I even went as far as to use [file-entry-cache](https://www.npmjs.com/package/file-entry-cache) directly. Ultimately, it was left out for being to specific to a non-transformative/single-step, file-based scenario. Instead, I provide an example of how one might implement a [gulp-based caching file-watch](https://github.com/adametry/gulp-eslint/blob/master/example/watch.js#L27) using `gulp-cached`.

**Comment by IanVS** [comment]
Excellent, I hadn't seen that, @adametry.  Thanks for pointing it out, I'll give it a try.  Closing this issue since it sounds like there are sufficient workarounds and it's not a great idea to put into `executeOnText`.
